Flaky TestDeltaLakeGcsConnectorSmokeTest.testOptimizeUsingForcedPartitioning

```
Error:  Tests run: 105, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 826.4 s <<< FAILURE! -- in io.trino.plugin.deltalake.TestDeltaLakeGcsConnectorSmokeTest
Error:  io.trino.plugin.deltalake.TestDeltaLakeGcsConnectorSmokeTest.testOptimizeUsingForcedPartitioning -- Time elapsed: 11.18 s <<< ERROR!
io.trino.testing.QueryFailedException: Failed to write Delta Lake transaction log entry
	at io.trino.testing.AbstractTestingTrinoClient.execute(AbstractTestingTrinoClient.java:133)
	at io.trino.testing.DistributedQueryRunner.executeWithQueryId(DistributedQueryRunner.java:505)
	at io.trino.testing.QueryAssertions.assertDistributedUpdate(QueryAssertions.java:107)
	at io.trino.testing.QueryAssertions.assertUpdate(QueryAssertions.java:61)
	at io.trino.testing.AbstractTestQueryFramework.assertUpdate(AbstractTestQueryFramework.java:424)
	at io.trino.testing.AbstractTestQueryFramework.assertUpdate(AbstractTestQueryFramework.java:419)
	at io.trino.plugin.deltalake.BaseDeltaLakeConnectorSmokeTest.testOptimizeUsingForcedPartitioning(BaseDeltaLakeConnectorSmokeTest.java:1961)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
	Suppressed: java.lang.Exception: SQL: INSERT INTO test_optimize_partitioned_table_13kkuabawe VALUES ('one', 6, 'test2', 9)
		at io.trino.testing.DistributedQueryRunner.executeWithQueryId(DistributedQueryRunner.java:509)
		... 12 more
Caused by: io.trino.spi.TrinoException: Failed to write Delta Lake transaction log entry
	at io.trino.plugin.deltalake.DeltaLakeMetadata.finishInsert(DeltaLakeMetadata.java:1912)
	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.finishInsert(ClassLoaderSafeConnectorMetadata.java:617)
	at io.trino.tracing.TracingConnectorMetadata.finishInsert(TracingConnectorMetadata.java:697)
	at io.trino.metadata.MetadataManager.finishInsert(MetadataManager.java:1140)
	at io.trino.tracing.TracingMetadata.finishInsert(TracingMetadata.java:694)
	at io.trino.sql.planner.LocalExecutionPlanner.lambda$createTableFinisher$4(LocalExecutionPlanner.java:4164)
	at io.trino.operator.TableFinishOperator.getOutput(TableFinishOperator.java:319)
	at io.trino.operator.Driver.processInternal(Driver.java:395)
	at io.trino.operator.Driver.lambda$process$8(Driver.java:298)
	at io.trino.operator.Driver.tryWithLock(Driver.java:694)
	at io.trino.operator.Driver.process(Driver.java:290)
	at io.trino.operator.Driver.processForDuration(Driver.java:261)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:887)
	at io.trino.execution.executor.timesharing.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:187)
	at io.trino.execution.executor.timesharing.TimeSharingTaskExecutor$TaskRunner.run(TimeSharingTaskExecutor.java:565)
	at io.trino.$gen.Trino_testversion____20231128_181115_1233.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.trino.spi.TrinoException: Failed to store statistics with table location: gs://trino-ci-test/test-delta-lake-integration-smoke-test-gu5i5s6vxd/test_optimize_partitioned_table_13kkuabawe
	at io.trino.plugin.deltalake.statistics.MetaDirStatisticsAccess.updateExtendedStatistics(MetaDirStatisticsAccess.java:109)
	at io.trino.plugin.deltalake.statistics.CachingExtendedStatisticsAccess.updateExtendedStatistics(CachingExtendedStatisticsAccess.java:74)
	at io.trino.plugin.deltalake.DeltaLakeMetadata.updateTableStatistics(DeltaLakeMetadata.java:3350)
	at io.trino.plugin.deltalake.DeltaLakeMetadata.finishInsert(DeltaLakeMetadata.java:1895)
	... 18 more
Caused by: java.io.IOException: Error writing file: gs://trino-ci-test/test-delta-lake-integration-smoke-test-gu5i5s6vxd/test_optimize_partitioned_table_13kkuabawe/_delta_log/_trino_meta/extended_stats.json
	at io.trino.filesystem.gcs.GcsUtils.handleGcsException(GcsUtils.java:39)
	at io.trino.filesystem.gcs.GcsOutputFile.createOutputStream(GcsOutputFile.java:94)
	at io.trino.filesystem.gcs.GcsOutputFile.createOrOverwrite(GcsOutputFile.java:63)
	at io.trino.filesystem.TrinoOutputFile.createOrOverwrite(TrinoOutputFile.java:35)
	at io.trino.filesystem.tracing.TracingOutputFile.lambda$createOrOverwrite$1(TracingOutputFile.java:57)
	at io.trino.filesystem.tracing.Tracing.withTracing(Tracing.java:47)
	at io.trino.filesystem.tracing.TracingOutputFile.createOrOverwrite(TracingOutputFile.java:57)
	at io.trino.plugin.deltalake.statistics.MetaDirStatisticsAccess.updateExtendedStatistics(MetaDirStatisticsAccess.java:98)
	... 21 more
Caused by: com.google.cloud.storage.StorageException: The object trino-ci-test/test-delta-lake-integration-smoke-test-gu5i5s6vxd/test_optimize_partitioned_table_13kkuabawe/_delta_log/_trino_meta/extended_stats.json exceeded the rate limit for object mutation operations (create, update, and delete). Please reduce your request rate. See https://cloud.google.com/storage/docs/gcs429.
	at com.google.cloud.storage.StorageException.translate(StorageException.java:170)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:313)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.create(HttpStorageRpc.java:392)
	at com.google.cloud.storage.StorageImpl.lambda$internalCreate$2(StorageImpl.java:213)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at com.google.cloud.storage.Retrying.run(Retrying.java:65)
	at com.google.cloud.storage.StorageImpl.run(StorageImpl.java:1514)
	at com.google.cloud.storage.StorageImpl.internalCreate(StorageImpl.java:210)
	at com.google.cloud.storage.StorageImpl.create(StorageImpl.java:142)
	at io.trino.filesystem.gcs.GcsOutputFile.createOutputStream(GcsOutputFile.java:84)
	... 27 more
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 429 Too Many Requests
POST https://storage.googleapis.com/upload/storage/v1/b/trino-ci-test/o?projection=full&uploadType=multipart
{
  "code" : 429,
  "errors" : [ {
    "domain" : "usageLimits",
    "message" : "The object trino-ci-test/test-delta-lake-integration-smoke-test-gu5i5s6vxd/test_optimize_partitioned_table_13kkuabawe/_delta_log/_trino_meta/extended_stats.json exceeded the rate limit for object mutation operations (create, update, and delete). Please reduce your request rate. See https://cloud.google.com/storage/docs/gcs429.",
    "reason" : "rateLimitExceeded"
  } ],
  "message" : "The object trino-ci-test/test-delta-lake-integration-smoke-test-gu5i5s6vxd/test_optimize_partitioned_table_13kkuabawe/_delta_log/_trino_meta/extended_stats.json exceeded the rate limit for object mutation operations (create, update, and delete). Please reduce your request rate. See https://cloud.google.com/storage/docs/gcs429."
}
	at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:146)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:118)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:37)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:570)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:493)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:603)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.create(HttpStorageRpc.java:389)
	... 36 more
```

https://github.com/trinodb/trino/actions/runs/7022450870/job/19107239986

--- a/lib/trino-filesystem-gcs/pom.xml
+++ b/lib/trino-filesystem-gcs/pom.xml
@@ -124,6 +124,11 @@
         </dependency>
 
         <dependency>
+            <groupId>org.threeten</groupId>
+            <artifactId>threetenbp</artifactId>
+        </dependency>
+
+        <dependency>
             <groupId>io.trino</groupId>
             <artifactId>trino-spi</artifactId>
             <scope>provided</scope>
@@ -138,6 +143,12 @@
         <dependency>
             <groupId>io.airlift</groupId>
             <artifactId>junit-extensions</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>io.airlift</groupId>
+            <artifactId>testing</artifactId>
             <scope>test</scope>
         </dependency>
 

--- a/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsFileSystemConfig.java
+++ b/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsFileSystemConfig.java
@@ -18,9 +18,14 @@ import io.airlift.configuration.ConfigDescription;
 import io.airlift.configuration.ConfigSecuritySensitive;
 import io.airlift.configuration.validation.FileExists;
 import io.airlift.units.DataSize;
+import io.airlift.units.Duration;
+import io.airlift.units.MinDuration;
 import jakarta.annotation.Nullable;
+import jakarta.validation.constraints.AssertTrue;
 import jakarta.validation.constraints.Min;
 import jakarta.validation.constraints.NotNull;
+
+import java.util.concurrent.TimeUnit;
 
 import static com.google.common.base.Preconditions.checkState;
 import static io.airlift.units.DataSize.Unit.MEGABYTE;
@@ -37,6 +42,12 @@ public class GcsFileSystemConfig
     private boolean useGcsAccessToken;
     private String jsonKey;
     private String jsonKeyFilePath;
+    private int maxRetries = 20;
+    private double backoffScaleFactor = 2.0;
+    private Duration maxRetryTime = new Duration(20, TimeUnit.SECONDS);
+    private Duration minBackoffDelay = new Duration(10, TimeUnit.MILLISECONDS);
+    // Note: there is no benefit to setting this much higher as the rpc quota is 1x per second: https://cloud.google.com/storage/docs/retry-strategy#java
+    private Duration maxBackoffDelay = new Duration(1100, TimeUnit.MILLISECONDS);
 
     @NotNull
     public DataSize getReadBlockSize()
@@ -146,6 +157,84 @@ public class GcsFileSystemConfig
     {
         this.jsonKeyFilePath = jsonKeyFilePath;
         return this;
+    }
+
+    @Min(0)
+    public int getMaxRetries()
+    {
+        return maxRetries;
+    }
+
+    @Config("gcs.client.max-retries")
+    @ConfigDescription("Maximum number of RPC attempts")
+    public GcsFileSystemConfig setMaxRetries(int maxRetries)
+    {
+        this.maxRetries = maxRetries;
+        return this;
+    }
+
+    @Min(1)
+    public double getBackoffScaleFactor()
+    {
+        return backoffScaleFactor;
+    }
+
+    @Config("gcs.client.backoff-scale-factor")
+    @ConfigDescription("Scale factor for RPC retry delay")
+    public GcsFileSystemConfig setBackoffScaleFactor(double backoffScaleFactor)
+    {
+        this.backoffScaleFactor = backoffScaleFactor;
+        return this;
+    }
+
+    @NotNull
+    public Duration getMaxRetryTime()
+    {
+        return maxRetryTime;
+    }
+
+    @Config("gcs.client.max-retry-time")
+    @ConfigDescription("Total time limit for an RPC to be retried")
+    public GcsFileSystemConfig setMaxRetryTime(Duration maxRetryTime)
+    {
+        this.maxRetryTime = maxRetryTime;
+        return this;
+    }
+
+    @NotNull
+    @MinDuration("0ms")
+    public Duration getMinBackoffDelay()
+    {
+        return minBackoffDelay;
+    }
+
+    @Config("gcs.client.min-backoff-delay")
+    @ConfigDescription("Minimum delay between RPC retries")
+    public GcsFileSystemConfig setMinBackoffDelay(Duration minBackoffDelay)
+    {
+        this.minBackoffDelay = minBackoffDelay;
+        return this;
+    }
+
+    @NotNull
+    @MinDuration("0ms")
+    public Duration getMaxBackoffDelay()
+    {
+        return maxBackoffDelay;
+    }
+
+    @Config("gcs.client.max-backoff-delay")
+    @ConfigDescription("Maximum delay between RPC retries.")
+    public GcsFileSystemConfig setMaxBackoffDelay(Duration maxBackoffDelay)
+    {
+        this.maxBackoffDelay = maxBackoffDelay;
+        return this;
+    }
+
+    @AssertTrue(message = "gcs.client.min-backoff-delay must be less than or equal to gcs.client.max-backoff-delay")
+    public boolean isRetryDelayValid()
+    {
+        return minBackoffDelay.compareTo(maxBackoffDelay) <= 0;
     }
 
     public void validate()

--- a/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsFileSystemConfig.java
+++ b/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsFileSystemConfig.java
@@ -150,12 +150,12 @@ public class GcsFileSystemConfig
 
     public void validate()
     {
-        // This cannot be normal validation, as it would make it impossible to write TestHiveGcsConfig.testExplicitPropertyMappings
+        // This cannot be normal validation, as it would make it impossible to write TestGcsFileSystemConfig.testExplicitPropertyMappings
 
         if (useGcsAccessToken) {
-            checkState(jsonKey == null, "Cannot specify 'hive.gcs.json-key' when 'hive.gcs.use-access-token' is set");
-            checkState(jsonKeyFilePath == null, "Cannot specify 'hive.gcs.json-key-file-path' when 'hive.gcs.use-access-token' is set");
+            checkState(jsonKey == null, "Cannot specify 'gcs.json-key' when 'gcs.use-access-token' is set");
+            checkState(jsonKeyFilePath == null, "Cannot specify 'gcs.json-key-file-path' when 'gcs.use-access-token' is set");
         }
-        checkState(jsonKey == null || jsonKeyFilePath == null, "'hive.gcs.json-key' and 'hive.gcs.json-key-file-path' cannot be both set");
+        checkState(jsonKey == null || jsonKeyFilePath == null, "'gcs.json-key' and 'gcs.json-key-file-path' cannot be both set");
     }
 }

--- a/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsStorageFactory.java
+++ b/lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsStorageFactory.java
@@ -13,6 +13,7 @@
  */
 package io.trino.filesystem.gcs;
 
+import com.google.api.gax.retrying.RetrySettings;
 import com.google.auth.oauth2.GoogleCredentials;
 import com.google.cloud.storage.Storage;
 import com.google.cloud.storage.StorageOptions;
@@ -20,6 +21,7 @@ import com.google.common.base.VerifyException;
 import com.google.common.collect.ImmutableList;
 import com.google.inject.Inject;
 import io.trino.spi.security.ConnectorIdentity;
+import org.threeten.bp.Duration;
 
 import java.io.ByteArrayInputStream;
 import java.io.FileInputStream;
@@ -29,6 +31,7 @@ import java.io.UncheckedIOException;
 import java.util.List;
 import java.util.Optional;
 
+import static com.google.cloud.storage.StorageRetryStrategy.getUniformStorageRetryStrategy;
 import static com.google.common.base.Strings.nullToEmpty;
 import static java.nio.charset.StandardCharsets.UTF_8;
 
@@ -39,6 +42,11 @@ public class GcsStorageFactory
     private final String projectId;
     private final boolean useGcsAccessToken;
     private final Optional<GoogleCredentials> jsonGoogleCredential;
+    private final int maxRetries;
+    private final double backoffScaleFactor;
+    private final Duration maxRetryTime;
+    private final Duration minBackoffDelay;
+    private final Duration maxBackoffDelay;
 
     @Inject
     public GcsStorageFactory(GcsFileSystemConfig config)
@@ -62,6 +70,12 @@ public class GcsStorageFactory
         else {
             jsonGoogleCredential = Optional.empty();
         }
+        this.maxRetries = config.getMaxRetries();
+        this.backoffScaleFactor = config.getBackoffScaleFactor();
+        // To avoid name collision by importing io.airlift.Duration
+        this.maxRetryTime = Duration.ofMillis(config.getMaxRetryTime().toMillis());
+        this.minBackoffDelay = Duration.ofMillis(config.getMinBackoffDelay().toMillis());
+        this.maxBackoffDelay = Duration.ofMillis(config.getMaxBackoffDelay().toMillis());
     }
 
     public Storage create(ConnectorIdentity identity)
@@ -81,7 +95,20 @@ public class GcsStorageFactory
             if (projectId != null) {
                 storageOptionsBuilder.setProjectId(projectId);
             }
-            return storageOptionsBuilder.setCredentials(credentials).build().getService();
+            // Note: without uniform strategy we cannot retry idempotent operations.
+            // The trino-filesystem api does not violate the conditions for idempotency, see https://cloud.google.com/storage/docs/retry-strategy#java for details.
+            return storageOptionsBuilder
+                    .setCredentials(credentials)
+                    .setStorageRetryStrategy(getUniformStorageRetryStrategy())
+                    .setRetrySettings(RetrySettings.newBuilder()
+                            .setMaxAttempts(maxRetries + 1)
+                            .setRetryDelayMultiplier(backoffScaleFactor)
+                            .setTotalTimeout(maxRetryTime)
+                            .setInitialRetryDelay(minBackoffDelay)
+                            .setMaxRetryDelay(maxBackoffDelay)
+                            .build())
+                    .build()
+                    .getService();
         }
         catch (IOException e) {
             throw new UncheckedIOException(e);

--- a/lib/trino-filesystem-gcs/src/test/java/io/trino/filesystem/gcs/TestGcsFileSystemConfig.java
+++ b/lib/trino-filesystem-gcs/src/test/java/io/trino/filesystem/gcs/TestGcsFileSystemConfig.java
@@ -15,6 +15,8 @@ package io.trino.filesystem.gcs;
 
 import com.google.common.collect.ImmutableMap;
 import io.airlift.units.DataSize;
+import io.airlift.units.Duration;
+import jakarta.validation.constraints.AssertTrue;
 import org.junit.jupiter.api.Test;
 
 import java.io.IOException;
@@ -25,6 +27,10 @@ import java.util.Map;
 import static io.airlift.configuration.testing.ConfigAssertions.assertFullMapping;
 import static io.airlift.configuration.testing.ConfigAssertions.assertRecordedDefaults;
 import static io.airlift.configuration.testing.ConfigAssertions.recordDefaults;
+import static io.airlift.testing.ValidationAssertions.assertFailsValidation;
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 public class TestGcsFileSystemConfig
@@ -33,14 +39,19 @@ public class TestGcsFileSystemConfig
     void testDefaults()
     {
         assertRecordedDefaults(recordDefaults(GcsFileSystemConfig.class)
-                .setReadBlockSize(DataSize.of(2, DataSize.Unit.MEGABYTE))
-                .setWriteBlockSize(DataSize.of(16, DataSize.Unit.MEGABYTE))
+                .setReadBlockSize(DataSize.of(2, MEGABYTE))
+                .setWriteBlockSize(DataSize.of(16, MEGABYTE))
                 .setPageSize(100)
                 .setBatchSize(100)
                 .setProjectId(null)
                 .setUseGcsAccessToken(false)
                 .setJsonKey(null)
-                .setJsonKeyFilePath(null));
+                .setJsonKeyFilePath(null)
+                .setMaxRetries(20)
+                .setBackoffScaleFactor(2.0)
+                .setMaxRetryTime(new Duration(20, SECONDS))
+                .setMinBackoffDelay(new Duration(10, MILLISECONDS))
+                .setMaxBackoffDelay(new Duration(1100, MILLISECONDS)));
     }
 
     @Test
@@ -58,17 +69,27 @@ public class TestGcsFileSystemConfig
                 .put("gcs.use-access-token", "true")
                 .put("gcs.json-key", "{}")
                 .put("gcs.json-key-file-path", jsonKeyFile.toString())
+                .put("gcs.client.max-retries", "10")
+                .put("gcs.client.backoff-scale-factor", "3.0")
+                .put("gcs.client.max-retry-time", "10s")
+                .put("gcs.client.min-backoff-delay", "20ms")
+                .put("gcs.client.max-backoff-delay", "20ms")
                 .buildOrThrow();
 
         GcsFileSystemConfig expected = new GcsFileSystemConfig()
-                .setReadBlockSize(DataSize.of(51, DataSize.Unit.MEGABYTE))
-                .setWriteBlockSize(DataSize.of(52, DataSize.Unit.MEGABYTE))
+                .setReadBlockSize(DataSize.of(51, MEGABYTE))
+                .setWriteBlockSize(DataSize.of(52, MEGABYTE))
                 .setPageSize(10)
                 .setBatchSize(11)
                 .setProjectId("project")
                 .setUseGcsAccessToken(true)
                 .setJsonKey("{}")
-                .setJsonKeyFilePath(jsonKeyFile.toString());
+                .setJsonKeyFilePath(jsonKeyFile.toString())
+                .setMaxRetries(10)
+                .setBackoffScaleFactor(3.0)
+                .setMaxRetryTime(new Duration(10, SECONDS))
+                .setMinBackoffDelay(new Duration(20, MILLISECONDS))
+                .setMaxBackoffDelay(new Duration(20, MILLISECONDS));
         assertFullMapping(properties, expected);
     }
 
@@ -95,5 +116,14 @@ public class TestGcsFileSystemConfig
                         .setJsonKeyFilePath("/dev/null")::validate)
                 .isInstanceOf(IllegalStateException.class)
                 .hasMessage("'gcs.json-key' and 'gcs.json-key-file-path' cannot be both set");
+
+        assertFailsValidation(
+                new GcsFileSystemConfig()
+                        .setJsonKey("{}")
+                        .setMinBackoffDelay(new Duration(20, MILLISECONDS))
+                        .setMaxBackoffDelay(new Duration(19, MILLISECONDS)),
+                "retryDelayValid",
+                "gcs.client.min-backoff-delay must be less than or equal to gcs.client.max-backoff-delay",
+                AssertTrue.class);
     }
 }

--- a/lib/trino-filesystem-gcs/src/test/java/io/trino/filesystem/gcs/TestGcsFileSystemConfig.java
+++ b/lib/trino-filesystem-gcs/src/test/java/io/trino/filesystem/gcs/TestGcsFileSystemConfig.java
@@ -80,20 +80,20 @@ public class TestGcsFileSystemConfig
                         .setUseGcsAccessToken(true)
                         .setJsonKey("{}}")::validate)
                 .isInstanceOf(IllegalStateException.class)
-                .hasMessage("Cannot specify 'hive.gcs.json-key' when 'hive.gcs.use-access-token' is set");
+                .hasMessage("Cannot specify 'gcs.json-key' when 'gcs.use-access-token' is set");
 
         assertThatThrownBy(
                 new GcsFileSystemConfig()
                         .setUseGcsAccessToken(true)
                         .setJsonKeyFilePath("/dev/null")::validate)
                 .isInstanceOf(IllegalStateException.class)
-                .hasMessage("Cannot specify 'hive.gcs.json-key-file-path' when 'hive.gcs.use-access-token' is set");
+                .hasMessage("Cannot specify 'gcs.json-key-file-path' when 'gcs.use-access-token' is set");
 
         assertThatThrownBy(
                 new GcsFileSystemConfig()
                         .setJsonKey("{}")
                         .setJsonKeyFilePath("/dev/null")::validate)
                 .isInstanceOf(IllegalStateException.class)
-                .hasMessage("'hive.gcs.json-key' and 'hive.gcs.json-key-file-path' cannot be both set");
+                .hasMessage("'gcs.json-key' and 'gcs.json-key-file-path' cannot be both set");
     }
 }

--- a/lib/trino-filesystem-gcs/src/test/java/io/trino/filesystem/gcs/TestGcsFileSystemGcs.java
+++ b/lib/trino-filesystem-gcs/src/test/java/io/trino/filesystem/gcs/TestGcsFileSystemGcs.java
@@ -13,10 +13,16 @@
  */
 package io.trino.filesystem.gcs;
 
+import io.trino.filesystem.TrinoOutputFile;
 import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.TestInstance;
 
 import java.io.IOException;
+import java.io.OutputStream;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.assertj.core.api.Assertions.assertThatNoException;
 
 @TestInstance(TestInstance.Lifecycle.PER_CLASS)
 public class TestGcsFileSystemGcs
@@ -27,5 +33,17 @@ public class TestGcsFileSystemGcs
             throws IOException
     {
         initialize(getRequiredEnvironmentVariable("GCP_CREDENTIALS_KEY"));
+    }
+
+    @Test
+    void testCreateFileRetry()
+    {
+        assertThatNoException().isThrownBy(() -> {
+            for (int i = 1; i <= 100; i++) {
+                TrinoOutputFile outputFile = getFileSystem().newOutputFile(getRootLocation().appendPath("testFile"));
+                try (OutputStream out = outputFile.createOrOverwrite()) {
+                    out.write("test".getBytes(UTF_8));
+                }
+            }});
     }
 }

--- a/plugin/trino-bigquery/pom.xml
+++ b/plugin/trino-bigquery/pom.xml
@@ -41,12 +41,6 @@
                 <artifactId>conscrypt-openjdk-uber</artifactId>
                 <version>2.5.2</version>
             </dependency>
-
-            <dependency>
-                <groupId>org.threeten</groupId>
-                <artifactId>threetenbp</artifactId>
-                <version>1.6.8</version>
-            </dependency>
         </dependencies>
     </dependencyManagement>
 

--- a/pom.xml
+++ b/pom.xml
@@ -2052,6 +2052,12 @@
             </dependency>
 
             <dependency>
+                <groupId>org.threeten</groupId>
+                <artifactId>threetenbp</artifactId>
+                <version>1.6.8</version>
+            </dependency>
+
+            <dependency>
                 <groupId>org.xerial.snappy</groupId>
                 <artifactId>snappy-java</artifactId>
                 <version>1.1.10.5</version>

